### DISCUSSION TITLE
Time to first response degraded by 2 -4 seconds since 0.4, SOLUTION

### DISCUSSION THREAD

**Discussion by kim-gtek** [open-webui/open-webui#7169] [General]
I noticed that OPENWEBUI takes 2-4 seconds longer to load chats.

By looking at the logs, the problem is clear, before every chat OPENWEBUI calls

 `get_all_models()` 

I have fixed this issue by adding the environment variables:
**backend/open_webui/config.py**
Line 533:
```
ALL_MODELS_CACHE_TTL = "1800"
OPENAI_MODELS_CACHE_TTL = "1800"
OLLAMA_MODELS_CACHE_TTL = "1800"
```

These are imported into the following files:

**backend/open_webui/apps/main.py**
Line 1007:
```
@cached(ttl=MODELS_CACHE_TTL) # Cache variable to control get_all_models()
async def get_all_models():
```

**backend/open_webui/apps/ollama/main.py**
Line 260:
```
@cached(ttl=OLLAMA_MODELS_CACHE_TTL)
async def get_all_models():
```

**backend/open_webui/apps/openai/main.py**
Line 372:
```
@cached(ttl=OPENAI_MODELS_CACHE_TTL)
async def get_all_models() -> dict[str, list]:
```

Obviously importing **from aiocache import cached** to each file, and adding the corresponding environment variable to **from open_webui.config import (**

**Comment by tkg61** [comment]
Seeing this as well!

**Comment by kim-gtek** [comment]
So I have changed my approach as I realized that my current workaround fixed the chat issue but cashed get_all_models when you add new connections or modify connections meaning you had to wait to see them.

My second fix is adding two **`get_all_model()`** passthroughs in **`backend/open_webui/apps/main.py`**

One for **`generate_chat_completions()`** to speed up chat generation, this one has the cache.
The other for **`get_models()`**

```
@cached(ttl=OPENAI_MODELS_CACHE_TTL)
async def get_all_models_chat_completion() -> dict[str, list]:
    return await get_all_models()

async def get_all_models_api_models() -> dict[str, list]:
    return await get_all_models()

async def get_all_models() -> dict[str, list]:
    log.info("get_all_models()")

    if not app.state.config.ENABLE_OPENAI_API:
        return {"data": []}

    responses = await get_all_models_responses()

    def extract_data(response):
        if response and "data" in response:
            return response["data"]
        if isinstance(response, list):
            return response
        return None

    models = {"data": merge_models_lists(map(extract_data, responses))}
    log.debug(f"models: {models}")

    return models


```

This fixes both problems in my opinion.

**Comment by tjbck** [comment]
Related: https://github.com/open-webui/open-webui/discussions/7060
